src/merkle_tree/constraints.rs

```diff
@@ -101,39 +45,166 @@ where
+                auth_path.push(TwoToOneH::OutputVar::new_variable(
+                    ark_relations::ns!(cs, "auth_path_node"),
+                    || Ok(v.clone()),
                     mode,
-                )?;
-                path.push((l_hash, r_hash));
+                )?)
             }
-            Ok(PathVar { path })
+            Ok(PathVar {
+                path,
+                auth_path,
+                leaf_sibling,
+                leaf_is_right_child: leaf_position_bit,
+            })
         })
     }
 }
 
+impl<P, LeafH, TwoToOneH, ConstraintF> PathVar<P, LeafH, TwoToOneH, ConstraintF>
+where
+    P: Config,
+    LeafH: CRHGadget<P::LeafHash, ConstraintF>,
+    TwoToOneH: TwoToOneCRHGadget<P::TwoToOneHash, ConstraintF>,
+    ConstraintF: Field,
+{
+    /// Given a leaf, calculate the root as if the merkle has this leaf at this path.
+    fn calculate_root(
+        &self,
+        leaf_hash_params: &LeafH::ParametersVar,
+        two_to_one_hash_params: &TwoToOneH::ParametersVar,
+        leaf: &impl ToBytesGadget<ConstraintF>,
+    ) -> Result<TwoToOneH::OutputVar, SynthesisError> {
+        let leaf_bytes = leaf.to_bytes()?;
+        let claimed_leaf_hash = LeafH::evaluate(leaf_hash_params, &leaf_bytes)?;
+        let leaf_sibling_hash = &self.leaf_sibling;
+
+        // calculate hash for the bottom non_leaf_layer
+
+        // At any given bit, the bit being 0 indicates our currently hashed value is the left,
+        // and the bit being 1 indicates our currently hashed value is on the right.
+        // Thus `left_hash` is sibling if leaf_position_bit is 1, and leaf if bit is 0
+
+        let left_hash = self
+            .leaf_is_right_child
+            .select(leaf_sibling_hash, &claimed_leaf_hash)?
+            .to_bytes()?;
+        let right_hash = self
+            .leaf_is_right_child
+            .select(&claimed_leaf_hash, leaf_sibling_hash)?
+            .to_bytes()?;
+
+        let mut curr_hash = TwoToOneH::evaluate(two_to_one_hash_params, &left_hash, &right_hash)?;
+        // To traverse up a MT, we iterate over the path from bottom to top (in reverse)
+
+        // At any given bit, the bit being 0 indicates our currently hashed value is the left,
+        // and the bit being 1 indicates our currently hashed value is on the right.
+        // Thus `left_hash` is the sibling if bit is 1, and it's the computed hash if bit is 0
+        for (bit, sibling) in self.path.iter().rev().zip(self.auth_path.iter().rev()) {
+            let left_hash = bit.select(sibling, &curr_hash)?;
+            let right_hash = bit.select(&curr_hash, sibling)?;
+
+            curr_hash = TwoToOneH::evaluate(
+                two_to_one_hash_params,
+                &left_hash.to_bytes()?,
+                &right_hash.to_bytes()?,
+            )?;
+        }
+
+        Ok(curr_hash)
+    }
+
+    /// Given a leaf, calculate the hash of the merkle tree
+    /// along the path and check if the given root is correct.
```

Comment:
```suggestion
    /// Check that hashing a Merkle tree path according to `self`, and
    /// with `leaf` as the leaf, leads to a Merkle tree root equalling `root`.
```

(Modify the above as you like)

Reply:
the change looks good.

Reply:
fixed